generator/common/beatgen/beatgen.go

```diff
@@ -0,0 +1,183 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package beatgen
+
+import (
+	"bufio"
+	"fmt"
+	"go/build"
+	"os"
+	"path/filepath"
+	"strings"
+
+	devtools "github.com/elastic/beats/dev-tools/mage"
+	"github.com/elastic/beats/generator/common/beatgen/setup"
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+// ConfigItem represents a value that must be configured for the custom beat
+type ConfigItem struct {
+	Key     string
+	Default func(map[string]string) string
+}
+
+// required user config for a custom beat
+// These are specified in env variables with newbeat_*
+var configList = []ConfigItem{
+	{
+		Key: "project_name",
+		Default: func(cfg map[string]string) string {
+			return "examplebeat"
+		},
+	},
+	{
+		Key: "github_name",
+		Default: func(cfg map[string]string) string {
+			return "your-github-name"
+		},
+	},
+	{
+		Key: "beat_path",
+		Default: func(cfg map[string]string) string {
+			ghName, _ := cfg["github_name"]
+			beatName, _ := cfg["project_name"]
+			return "github.com/" + ghName + "/" + strings.ToLower(beatName)
+		},
+	},
+	{
+		Key: "full_name",
+		Default: func(cfg map[string]string) string {
+			return "Firstname Lastname"
+		},
+	},
+	{
+		Key: "type",
+		Default: func(cfg map[string]string) string {
+			return "beat"
+		},
+	},
+}
+
+// Generate generates a new custom beat
+func Generate() error {
+	cfg, err := getConfig()
+	if err != nil {
+		return err
+	}
+	err = setup.GenNewBeat(cfg)
+	if err != nil {
+		return err
+	}
+
+	absBeatPath := filepath.Join(build.Default.GOPATH, "src", cfg["beat_path"])
+
+	err = os.Chdir(absBeatPath)
+	if err != nil {
+		return err
+	}
+
+	mg.Deps(setup.CopyVendor)
+	mg.Deps(setup.RunSetup)
+	mg.Deps(setup.GitInit)
+
+	if cfg["type"] == "metricbeat" {
+		//This is runV because it'll ask for user input, so we need stdout.
+		err = sh.RunV("make", "create-metricset")
+		if err != nil {
+			return errors.Wrap(err, "error running create-metricset")
+		}
+	}
+
+	mg.Deps(setup.Update)
+	mg.Deps(setup.GitAdd)
+
+	return nil
+}
+
+// VendorUpdate updates the beat vendor directory
+func VendorUpdate() error {
+	err := sh.Rm("./vendor/github.com/elastic/beats")
+	if err != nil {
+		return errors.Wrap(err, "error removing vendor dir")
+	}
+
+	devtools.SetElasticBeatsDir(getAbsoluteBeatsPath())
+	mg.Deps(setup.CopyVendor)
+	return nil
+}
+
+// returns a "compleated" config object with everything we need
+func getConfig() (map[string]string, error) {
+	userCfg := make(map[string]string)
+	for _, cfgVal := range configList {
+		var cfgKey string
+		var err error
+		cfgKey, isSet := getEnvConfig(cfgVal.Key)
+		if !isSet {
+			cfgKey, err = getValFromUser(cfgVal.Key, cfgVal.Default(userCfg))
+			if err != nil {
+				return userCfg, err
+			}
+		}
+		userCfg[cfgVal.Key] = cfgKey
+	}
+
+	return userCfg, nil
+
+}
+
+func getEnvConfig(cfgKey string) (string, bool) {
```

Comment:
This function has a strong package dependency on `creator.go` in `github.com/elastic/beats/generator/common/beatgen/setup`